In this step data is to be understood more deeply before applying methods to clean the data

# Types of unclean data

## 1. Dirty Data

  #### a. Data with quality isssues     
  ####  b. Duplicate Data
  ####   c. Missing Values
  ####  d. Corrupt data
  ####   e. Innaccurate data

## 2. Messy Data

 when the data has structural issues

# Steps in Assessment

### 1.Discover    
### 2. Document

# Data Quality Dimensions

- 1.Completeness issue
- 2.Validity issue (eg. height is -ve)
- 3.Accuracy issue (eg. height is 2cm)
- 4.Consistency issues (eg. New York sometimes refered as NY and other times as full name) 

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse = pd.read_csv('adverse_reactions.csv')

In [3]:
print(patients.shape)
print(treatments.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


In [7]:
patients.head(3)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2       402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8

In [8]:
treatments.head(3)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN

In [9]:
adverse.head(3)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia

In [13]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [14]:
patients[patients['city'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [15]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [16]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [17]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [19]:
#gives randomly selected rows
patients.sample(3)

patient_id assigned_sex given_name surname               address  \
83           84         male       Chen     Yao    2566 Ingram Street   
489         490       female    Jasmine   Sykes     2607 Water Street   
199         200         male     Zdeněk   Synek  3818 Strother Street   

          city       state  zip_code        country  \
83      Dayton          OH   45407.0  United States   
489  Lafayette  California   94549.0  United States   
199       Dora          AL   35062.0  United States   

                                      contact  birthdate  weight  height   bmi  
83            937-523-1224ChenYao@teleworm.us  11/8/1975   214.7      76  26.1  
489   JasmineSykes@jourrapide.com925-283-5425  12/1/1988   187.2      63  33.2  
199  ZdenekSynek@jourrapide.com1 205 648 4197  9/27/1950   168.3      70  24.1

In [22]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

## Quality

[Completeness(C), Validity(V), Accuracy(A), Consistency(Co)]

#### 1.Patients

- zipcode column is stored in decimal format [V]
- certain zipcodes are of four digits  [V]
- for the state column sometimes full names are used and other times short forms are used [Co]
- missing values of contact details of 12 patients [C]
- incorrect data types for assigned_sex and birthdate columns [??]
- Errorneous data for weights and heights [A]
- contact column has inconsistency in format of data [Co]
- duplicate and useless data in dataset [Co]

#### 2.Treatments :

- hba1c_change column has missing values [C]
- novodra and auralin columns has 'u' attatched to numbers [V]
- only 280 rows are there instead of 350 [C]
- name and surname column are in lower case [Co]
- auralin and novodra have incorrect data type [V]
- hba1c_change column change 9 to 4 [A]
- use of dashes instead of nan in auralin and novodra columns [Co]

#### 3. Adverse Effects

- name columns are in lower case [Co]

## Tidiness

#### 1.Patients

- the contact column has phone and email together

#### 2.Treatments

- three variables stored in 2 columns auralin and novodra columns

#### 3. Adverse Effects

- no point of using table separately we should fuse it into treatment table

# DATA CLEANING

### Follow the order provided below - 
- Completeness issues
- Tidiness issues
- Remaining issues such as validity , accuuracy and consistency

### Steps involved in cleaning data - 
- Define
- Code 
- Test

#### always make a copy of pandas df before cleaning process

In [4]:
patients_copy = patients.copy()
treatments_copy = treatments.copy()
adverse_copy = adverse.copy()

## a) Completeness Issues
- missing values of contact details of 12 patients in patients table
- only 280 rows are there instead of 350 in treatments table
- hba1c_change column has missing values in treatments table

In [8]:
# we can't do anything for first two so we will handle the last one

### Define

In [9]:
# Subtract hba1c_end from hba1c_start to get values of hba1c_change

### Code

In [5]:
treatments_copy['hba1c_change'] = treatments_copy['hba1c_start'] - treatments_copy['hba1c_end']

### Test

In [6]:
treatments_copy['hba1c_change'].isnull().sum()

0

no null values left in this column

## b) Tidiness Issues

#### 1.Patients

- the contact column has phone and email together

#### 2.Treatments

- three variables stored in 2 columns auralin and novodra columns

#### 3. Adverse Effects

- no point of using table separately we should fuse it into treatment table

## Define

In [ ]:
# 2. three variables stored in 2 columns auralin and novodra columns
# we will use the melt function on these columns

## Code

In [7]:
pd.melt(treatments_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], var_name='treatment', value_name='dose')

given_name       surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika      jindrová         7.63       7.20          0.43   auralin   
1       elliot    richardson         7.56       7.09          0.47   auralin   
2     yukitaka      takenaka         7.68       7.25          0.43   auralin   
3         skye    gormanston         7.97       7.62          0.35   auralin   
4       alissa        montez         7.78       7.46          0.32   auralin   
..         ...           ...          ...        ...           ...       ...   
555     albina      zetticci         7.93       7.73          0.20   novodra   
556       john   teichelmann         7.90       7.58          0.32   novodra   
557     mathea       lillebø         9.04       8.67          0.37   novodra   
558     vallie        prince         7.64       7.28          0.36   novodra   
559     samúel  guðbrandsson         8.00       7.64          0.36   novodra   

          dose  
0    41u - 48u  
1            -  
2            -  
3    33u - 36u  
4            -  
..         ...  
555          -  
556  49u - 49u  
557          -  
558          -  
559          -  

[560 rows x 7 columns]

In [8]:
treatments_copy = pd.melt(treatments_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], var_name='treatment', value_name='dose')

In [11]:
treatments_copy = treatments_copy[treatments_copy['dose'] != '-']

In [12]:
treatments_copy

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika     jindrová         7.63       7.20          0.43   auralin   
3         skye   gormanston         7.97       7.62          0.35   auralin   
6       sophia       haugen         7.65       7.27          0.38   auralin   
7        eddie       archer         7.89       7.55          0.34   auralin   
9         asia      woźniak         7.76       7.37          0.39   auralin   
..         ...          ...          ...        ...           ...       ...   
548     yasmin       araujo         7.82       7.36          0.46   novodra   
549     hiromu     horikawa         7.77       7.28          0.49   novodra   
551        leo       vieira         7.74       7.36          0.38   novodra   
552     steven          roy         7.87       7.43          0.44   novodra   
556       john  teichelmann         7.90       7.58          0.32   novodra   

          dose  
0    41u - 48u  
3    33u - 36u  
6    37u - 42u  
7    31u - 38u  
9    30u - 36u  
..         ...  
548  51u - 54u  
549  47u - 46u  
551  30u - 33u  
552  41u - 43u  
556  49u - 49u  

[280 rows x 7 columns]

In [35]:
treatments_copy['lower_dose'] = treatments_copy['dose'].str.split('-').apply(lambda x:x[0])
treatments_copy['higher_dose'] = treatments_copy['dose'].str.split('-').apply(lambda x:x[1])

In [40]:
treatments_copy.drop(columns=['dose'], inplace=True)

## Test

In [42]:
treatments_copy

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika     jindrová         7.63       7.20          0.43   auralin   
3         skye   gormanston         7.97       7.62          0.35   auralin   
6       sophia       haugen         7.65       7.27          0.38   auralin   
7        eddie       archer         7.89       7.55          0.34   auralin   
9         asia      woźniak         7.76       7.37          0.39   auralin   
..         ...          ...          ...        ...           ...       ...   
548     yasmin       araujo         7.82       7.36          0.46   novodra   
549     hiromu     horikawa         7.77       7.28          0.49   novodra   
551        leo       vieira         7.74       7.36          0.38   novodra   
552     steven          roy         7.87       7.43          0.44   novodra   
556       john  teichelmann         7.90       7.58          0.32   novodra   

    lower_dose higher_dose  
0         41u          48u  
3         33u          36u  
6         37u          42u  
7         31u          38u  
9         30u          36u  
..         ...         ...  
548       51u          54u  
549       47u          46u  
551       30u          33u  
552       41u          43u  
556       49u          49u  

[280 rows x 8 columns]

## Define

In [43]:
# lets merge the adverse effects table into treatments

## Code

In [65]:
new_df = pd.merge(treatments_copy, adverse_copy, how='outer', left_on=['given_name','surname'], right_on=['given_name','surname'])

In [66]:
new_df

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
1           skye  gormanston         7.97       7.62          0.35   auralin   
2         sophia      haugen         7.65       7.27          0.38   auralin   
3          eddie      archer         7.89       7.55          0.34   auralin   
4           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
283        sofia   hermansen          NaN        NaN           NaN       NaN   
284        jakob    jakobsen          NaN        NaN           NaN       NaN   
285  christopher    woodward          NaN        NaN           NaN       NaN   
286     miłosław  wiśniewski          NaN        NaN           NaN       NaN   
287        lixue       hsueh          NaN        NaN           NaN       NaN   

    lower_dose higher_dose           adverse_reaction  
0         41u          48u                        NaN  
1         33u          36u                        NaN  
2         37u          42u                        NaN  
3         31u          38u                        NaN  
4         30u          36u                        NaN  
..         ...         ...                        ...  
283        NaN         NaN  injection site discomfort  
284        NaN         NaN               hypoglycemia  
285        NaN         NaN                     nausea  
286        NaN         NaN  injection site discomfort  
287        NaN         NaN  injection site discomfort  

[288 rows x 9 columns]

In [70]:
new_df[new_df['adverse_reaction'].notnull()]

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
5         joseph         day         7.70       7.19          0.51   auralin   
6         joseph         day         7.70       7.19          0.51   auralin   
14       clinton      miller         7.79       7.40          0.39   auralin   
34        louise     johnson         7.63       7.32          0.31   auralin   
43        finley    chandler         7.65       7.26          0.39   auralin   
47        idalia       moore         8.34       7.90          0.44   auralin   
52        xiuxiu       chang         7.64       7.23          0.41   auralin   
86      gabriele     saenger         8.35       7.97          0.38   auralin   
93         merci      leroux         8.98       8.64          0.34   auralin   
111     anenechi       chidi         7.64       7.31          0.33   auralin   
119      manouck     wubbels         7.66       7.40          0.26   auralin   
125       jia li        teng         7.66       7.32          0.34   auralin   
127         alex    crawford         7.69       7.30          0.39   auralin   
137       elliot  richardson         7.56       7.09          0.47   novodra   
140      jasmine       sykes         7.56       7.18          0.38   novodra   
156    krisztina      magyar         7.52       7.01          0.51   novodra   
164      albinca     komavec         7.89       7.46          0.43   novodra   
175          ole    petersen         7.95       7.55          0.40   novodra   
187    abdul-nur         isa         7.98       7.53          0.45   novodra   
189         abel     yonatan         7.88       7.50          0.38   novodra   
226        tegan     johnson         7.79       7.43          0.36   novodra   
232       monika      lončar         9.13       8.72          0.41   novodra   
237         lena        baer         7.70       7.40          0.30   novodra   
243      cecilie      nilsen         9.63       9.21          0.42   novodra   
262         kang         mai         7.78       7.45          0.33   novodra   
274         leon      scholz         7.72       7.29          0.43   novodra   
278       steven         roy         7.87       7.43          0.44   novodra   
280        berta  napolitani          NaN        NaN           NaN       NaN   
281       flavia  fiorentino          NaN        NaN           NaN       NaN   
282          noe      aranda          NaN        NaN           NaN       NaN   
283        sofia   hermansen          NaN        NaN           NaN       NaN   
284        jakob    jakobsen          NaN        NaN           NaN       NaN   
285  christopher    woodward          NaN        NaN           NaN       NaN   
286     miłosław  wiśniewski          NaN        NaN           NaN       NaN   
287        lixue       hsueh          NaN        NaN           NaN       NaN   

    lower_dose higher_dose           adverse_reaction  
5         29u          36u               hypoglycemia  
6         29u          36u               hypoglycemia  
14        42u          51u          throat irritation  
34        32u          42u               hypoglycemia  
43        31u          45u                   headache  
47        25u          31u               hypoglycemia  
52        32u          41u               hypoglycemia  
86        22u          30u               hypoglycemia  
93        27u          33u               hypoglycemia  
111       52u          61u               hypoglycemia  
119       55u          62u          throat irritation  
125       48u          54u                     nausea  
127       51u          62u               hypoglycemia  
137       40u          45u               hypoglycemia  
140       42u          44u               hypoglycemia  
156       32u          27u               hypoglycemia  
164       41u          39u               hypoglycemia  
175       29u          32u               hypoglycemia  
187       54u          50u               hypoglycemia  
189  

In [ ]:
new_df.drop(new)